File: clients/java/src/main/java/org/aigen/oabp/model/Currency.java
package org.aigen.oabp.model;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

/**
 * The settlement unit of a mission reward.
 *
 * <ul>
 *   <li>{@link #AIGEN} — the protocol's uncapped, off-chain reputation/points token.</li>
 *   <li>{@link #USDC} — a real-value stablecoin payout.</li>
 * </ul>
 *
 * <p>Deserialization is tolerant of unknown/extra currencies the server may add later:
 * any unrecognized value maps to {@link #UNKNOWN} rather than throwing, so a new
 * server-side currency never breaks an older client. {@link #UNKNOWN} is never sent
 * back to the server (it has no wire value of its own — see {@link #wireValue()}).
 */
public enum Currency {
    AIGEN("AIGEN"),
    USDC("USDC"),
    /** Sentinel for a currency the server returned that this SDK version does not know. */
    UNKNOWN(null);

    private final String wire;

    Currency(String wire) {
        this.wire = wire;
    }

    /**
     * @return the exact token used on the wire, e.g. {@code "AIGEN"}.
     * @throws IllegalStateException if called on {@link #UNKNOWN}, which has no wire form.
     */
    @JsonValue
    public String wireValue() {
        if (wire == null) {
            throw new IllegalStateException("Currency.UNKNOWN has no wire representation");
        }
        return wire;
    }

    /**
     * Parses a wire string, case-insensitively, never throwing on unknown input.
     *
     * @param value raw currency string from JSON, may be {@code null}
     * @return the matching constant, or {@link #UNKNOWN} if unrecognized/null
     */
    @JsonCreator
    public static Currency fromWire(String value) {
        if (value == null) {
            return UNKNOWN;
        }
        for (Currency c : values()) {
            if (c.wire != null && c.wire.equalsIgnoreCase(value)) {
                return c;
            }
        }
        return UNKNOWN;
    }
}

File: clients/kotlin/src/main/kotlin/org/aigen/oabp/model/Currency.kt
package org.aigen.oabp.model

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

/**
 * The settlement unit of a mission reward.
 *
 *  - [AIGEN] — the protocol's uncapped, off-chain reputation/points token.
 *  - [USDC] — a real-value stablecoin payout.
 *
 * Deserialization is tolerant of currencies the server may add later: any unrecognized
 * value maps to [UNKNOWN] rather than throwing, so a new server-side currency never breaks
 * an older client. [UNKNOWN] has no wire form and must never be sent back to the server
 * (see [wireValue]).
 */
@Serializable(with = CurrencySerializer::class)
public enum class Currency(
    public val wire: String?,
) {
    AIGEN("AIGEN"),
    USDC("USDC"),

    /** Sentinel for a currency the server returned that this SDK version does not know. */
    UNKNOWN(null),
    ;

    /**
     * The exact token used on the wire, e.g. `"AIGEN"`.
     *
     * @throws IllegalStateException if called on [UNKNOWN], which has no wire form.
     */
    public fun wireValue(): String = wire ?: error("Currency.UNKNOWN has no wire representation")

    public companion object {
        /**
         * Parses a wire string case-insensitively, never throwing on unknown input.
         *
         * @return the matching constant, or [UNKNOWN] if unrecognized/null.
         */
        public fun fromWire(value: String?): Currency {
            if (value == null) return UNKNOWN
            return entries.firstOrNull { it.wire != null && it.wire.equals(value, ignoreCase = true) }
                ?: UNKNOWN
        }
    }
}

/** kotlinx.serialization adapter: encodes [Currency.wireValue], decodes via [Currency.fromWire]. */
public object CurrencySerializer : KSerializer<Currency> {
    override val descriptor: SerialDescriptor =
        PrimitiveSerialDescriptor("org.aigen.oabp.Currency", PrimitiveKind.STRING)

    override fun serialize(
        encoder: Encoder,
        value: Currency,
    ) {
        encoder.encodeString(value.wireValue())
    }

    override fun deserialize(decoder: Decoder): Currency = Currency.fromWire(decoder.decodeString())
}
